File: LeetCode/94二叉树的中序遍历.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void inOrder(TreeNode root, List<Integer> ans) {
        if (root == null) return ;
        inOrder(root.left, ans);
        ans.add(root.val);
        inOrder(root.right, ans);
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        inOrder(root, ans);
        return ans;
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        Stack<TreeNode> sta = new Stack<>();
        TreeNode cur = root;
        while (cur != null || !sta.isEmpty()) {
            if (cur != null) {
                sta.push(cur);
                cur = cur.left;
            } else {
                cur = sta.pop();
                ans.add(cur.val);
                cur = cur.right;
            }
        }
        return ans;
    }
}

File: LeetCode/94二叉树的中序遍历.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inOrder(root *TreeNode, ans *[]int) {
	if root == nil {
		return
	}
	inOrder(root.Left, ans)
	*ans = append(*ans, root.Val)
	inOrder(root.Right, ans)
}

func inorderTraversal(root *TreeNode) (ans []int) {
	inOrder(root, &ans)
	return ans
}

/**
 * 用栈递归
 */
func inorderTraversal(root *TreeNode) (ans []int) {
	sta := []*TreeNode{}
	if root == nil {
		return
	}
	for root != nil || len(sta) > 0 {
		if root != nil {
			sta = append(sta, root)
			root = root.Left
		} else {
			root = sta[len(sta)-1]
			sta = sta[:len(sta)-1]
			ans = append(ans, root.Val)
			root = root.Right
		}
	}
	return
}